src/java/org/apache/cassandra/io/util/CompressedChunkReader.java

```diff
@@ -117,8 +141,23 @@ public void readChunk(long position, ByteBuffer uncompressed)
                 {
                     ByteBuffer compressed = bufferHolder.getBuffer(length);
 
-                    if (channel.read(compressed, chunk.offset) != length)
-                        throw new CorruptBlockException(channel.filePath(), chunk);
+                    if (readAheadBuffer != null && readAheadBuffer.hasBuffer())
+                    {
+                        int copied = 0;
+                        while (copied < length) {
```

Comment:
if I do the following test everything should blow up

1) read ahead of 4k
2) chunk size of 64k

I don't see any place where you check that the amount of data to read is *safe* to read?

Reply:
There need to be configuration checks to make sure that read ahead is either disabled or some minimum multiple of chunk size, right? Even if your situation could be made to not explode, it’s non-sensical?

Reply:
that brings up an interesting point in terms of validation. the read ahead buffer size is a cassandra.yaml setting but chunk size is table level. Some thoughts/questions:
    
* Do we make the buffer size a table level setting with a default in the YAML? I don't love it as a table level parameter because its more driven by the disks than the data shape. 
* if we leave it how it is what do we do if someone creates or modifies a table such that chunk size > buffer size. This should be rare because buffer size should be 64k or greate but still we have to handle it. One option I see is to log and warn. 

Thoughts @maedhroz @dcapwell @rustyrazorblade

Reply:
-1 to making it a table level option.  No benefit and adds unnecessary complexity.

I think the problem here stems from allowing the user to specify a small buffer size.  It doesn't make sense to use a buffer less than 256KB.  This isn't just because of the way EBS works (mentioned in the ticket) but because of the way *every* disk works.  

The buffer should be > chunk size but also >= 256 and using less is an unnecessary configuration option that introduces both edge cases and sub-optimal performance.

Reply:
@rustyrazorblade @maedhroz would you prefer a boolean to turn it on and off and a fixed buffer size or validation to ensure its >= 256? if we go with the the latter we still have the chunk size / buffer size validation challenge because the max chunk size is 2GB

Reply:
If someone is using a massive chunk length (I've never seen this in 12 years) then it doesn't really matter if we have the buffer in place because the reads will be fairly large anyways.  So I think whatever is easiest for you to add is fine.  Can you just bypass it in that case?

Reply:
Ya unfortunately whether people use it, which would be surprising, we allow it so we have to handle it. I do like having it be a size. I also like ensuring it’s not smaller than some value. And I think I agree if the buffer size is < chunk size just don’t take the new code path (as chunk size will be sufficient anyways since it’s so large in that case). Only open question then is do we want to log when it happens? And do we log when the table metadata is changed or when the scanner is created (I lean towards the former)?

Reply:
> would you prefer a boolean to turn it on and off and a fixed buffer size.

I think it could be useful to test larger sizes.  We might find that a 1MB buffer works a little better in some situations, such as spinning disks or a slower NAS / SAN.

Reply:
I also made it a hot property